Opening sections for the Standards (#8)
uptane/uptane-standard: uptane-standard.md

in terms of memory, storage space, connection to the internet, and 
access to a time source. Primary units are updated
directly, while secondary units are generally updated by a primary.

*Repository*: Designated server containing images and metadata.

*Suppliers*: Indepedent companies to which auto manufacturers may 
outsource the production of ECUs. Tier-1 suppliers directly serve the
manufacturers. Tier-2 suppliers are those that receive outsourced work
from Tier-1 suppliers.

*Telematics Systems*: Electronic system that can remotely monitor a vehicle
or its individual components.

*Vehicle Version Manifest*: A compliation of manifests from all ECUs on
a vehicle. It serves as a master list of all programs currretly
running on alll ECUs in the vehcile.

## **2.3. Uptane Role Terminology**

These terms are defined in greater detail in Section 5.

*Delegations*: Designating the responsibility of signing metadata to another entity. 

*Roles*: The roles mechanism of Uptane allows
the system to distribute signing responsibilities so compromise
of one key does not impact other parts of the system.

  * *Root Role*: Distributes and revokes public keys used to 

>   verify the root, timestamp, release, and targets role metadata.

Author: Damn. I fixed the snapshot reference elsewhere, but missed this one.

*ECUs*: Electronic Control Units, the computing units on vehicle.

*LIN Bus*: Local Interconnect Bus

*SOTA*: Software Over-the-Air

*VIN*: Vehicle identification Number

# **3. Rationale for and Scope of Uptane Standards**
This Standards document clarifies the essential components and best practices 
for the secure design implentation and deployment of Uptane by OEMs and suppiers.
These practices contribute to compromise resilience, or the ability
to minimize the extent of the threat posed by any given attack.

## **3.1. Why Uptane requires standards**
A standards document that can guide the safe design, integration and deployment of
Uptane in cars is needed at this time because:

* The number of connected units on the average vehicle continues to grow, with 
mainstream cars now containing up to [10 million lines](https://www.usatoday.com/story/tech/columnist/2016/06/28/your-average-car-lot-more-code-driven-than-you-think/86437052/)
of code.

* The [expanded use of software over-the-air](https://www.consumerreports.org/automotive-technology/automakers-embrace-over-the-air-updates-can-we-trust-digital-car-repair/)
strategies creates new attack surfaces for malicious parties.

> * Legacy update strategies, such as SSL/TLS or GPG/RSA, are not feasible for use

Author: This might not even be what belongs here. I got a little confused because in the sample standard that Ira was nice enough to pass along to me---because I had no idea what this was supposed to address--was providing a rationale for the standard and not the technology. So, I was sort of splitting the difference here in making a case for both Uptane and the new standard. As a result it may be too muddled, but I figured I'd let someone who knows better than me resolve what information this section needed.

* Physical attacks, such as manual tampering with ECUs outside the
vehicle.

* Compromise of the supply chain (e.g., build system, version control system, 
packaging process). A number of strategies already (e.g., git signing, TPMs, in-toto)
exist to address this problem. Therefore, there is no need duplicate those
techniques here. 

* Problems associated with OBD or UDS programming of ECUs, such as
authentication of communications between ECUs.

## **3.2.4 Design Requirements**

The design requirements for this document are goverened by three principal
parameters: 

* to clearly mandate the design and implementation steps that are 
security critical and must be followed as is, while offering flexibility
in the implementation of non-critical steps. In this manner, users 
can adapt to support different use models and deployment scenarios. 

* to delineate best practices to ensure that, should a vehicle 
be attacked, an attacker is forced to compromise many different
systems.

* to ensure that, if implemented, the security practices mandated or 
suggested in this document do not interfere with the functionality
of ECUs, vehicles, or the

> manufacturing systems on which they run.

Reviewer: Let's just say "software repositories"? :)

> manufacturing systems on which they run.

Reviewer: ```suggestion
software repositories on which they run.
```

number of strategies, including the following.

  * *Drop-request attack:* blocks network traffic outside or inside the 
  vehicle.
  * *Slow retrieval attack:* slows down delivery of updates to ECUs so a 
  known security vulnerability can be exploited before a corrective patch
  is received.
  * *Freeze attack:* continues to send the last known update to an ECU, 
  even if a newer update exists.
  * *Partial bundle installation attack:* drips traffic to selected 
  Allows only part of an update
  to install by dropping traffic to selected ECUs.
  
## **4.3. Interfere with ECU functionality:** 
Attackers seeking to change the actual
functionality of vehicle ECUs may do so in one of the following ways:

  * *Rollback attack:* tricks an ECU into installing outdated software with
  known vulnerabilities.
  * Endless data attack: causes an ECU to crash by sending it an infinite amount
  of data until it runs out of storage.
  * *Mixed-bundles attack:* shuts down an ECU by causing it to install incompatible
  versions of software updates that must not be installed at the same time. 
  Attackers can accomplish this by showing different bundles to different 
  ECUs at the same time.

>   * *Mix-and-match attack:* If attackers have compromised repository keys, 

Reviewer: ```suggestion
  * *Mix-and-match attack:* release arbitrary combinations of new versions of images that were not meant to be installed together, given that attackers have compromised repository keys.
```

>   they can use these keys to release arbitrary combinations of new versions of images.

Reviewer: ```suggestion
```

  
## **4.4. Control the ECU:**
The last and most severe threat is if an attack seeks to remotely control the ECU.
The attacker can modify the vehicle’s performance through an arbitrary software
attack, in which the software on an ECU is overwritten with a malicious software program.

# **5. Detailed Design of Uptane**

At a high level, Uptane requires:

* Two software repositories:

>     * An image repository containing binary images for install, and signed metadata about those images

Reviewer: ```suggestion
    * An image repository containing binary images for install, and signed metadata about those images.
```

>     * A director repository connected to an inventory database that can sign metadata on demand for images

Reviewer: We should define "inventory database"

>     in the image repository

Reviewer: ```suggestion
    on the image repository.
```

* Repository tools for generating Uptane-specific metadata about images
* A public key infrastructure supporting the required metadata production/signing 
roles on each repository:
    * Root - Certificate authority for the repo. Distributes public keys for verifying

>     all the other roles' metadata

Reviewer: ```suggestion
    all the other roles' metadata.
```

    * Timestamp - Indicates whether there are new metadata or images
    * Snapshot - Indicates images released by the repository at a point in time, 

>     via signing metadata about targets metadata

Reviewer: ```suggestion
    via signing metadata about targets metadata.
```

    * Targets - Indicates metadata about images, such as hashes and file sizes
* A time server to deliver cryptographically verifiable time to ECUs
* An in-vehicle client on a primary ECU capable of verifying the signatures on all 
update metadata, handling all server communication, and downloading updates on

> behalf of secondary ECUs

Reviewer: ```suggestion
behalf of secondary ECUs.
```

* A client or library on each secondary ECU capable of performing either

> full or partial verification of metadata

Reviewer: ```suggestion
full or partial verification of metadata.
```

> full or partial verification of metadata

Reviewer: Have we defined full / partial verification?

## **5.1. Roles on repositories {#roles}**

A repository contains images and metadata. Each role has a particular type of 
metadata associated with it, as described in {{meta_syntax}}.

### **5.1.1. The Root role {#root_role}**

The Root role SHALL be responsible for a Certificate Authority as defined in {{RFC3647}}.
The Root role SHALL produce and sign Root metadata as described in {{root_meta}}.
The Root role SHALL sign the public keys used to verfy the metadata produced by the

> Timestamp, Snapshot, and Targets roles.

Reviewer: ```suggestion
Root, Timestamp, Snapshot, and Targets roles.
```

> The Root role SHALL revoke keys for the other roles, in case of compromise.

Reviewer: ```suggestion
The Root role SHALL revoke and replace keys for all of these roles, in case of compromise.
```

### **5.1.2. The Targets role {#targets_role}**

The Targets role SHALL produce and sign metadata about images and delegations as described in {{targets_meta}}.

#### **5.1.2.1. Delegations**

The Targets role on the Image repository MAY delegate the responsibility of signing metadata
to other, custom-defined roles. If it does, it MUST do so as specified in {{delegations_meta}}.

Responsibility for signing images or a subset of images MAY be delegated to more than one
role, and therefore it is possible for two different roles to be trusted for signing a
particular image. For this reason, delegations MUST be prioritized.

A particular delegation for a subset of images MAY be designated as **terminating**.
For terminating delegations, the client SHALL NOT search the any further if it does
not find validly signed metadata about those images in the terminating delegation. 
Delegations SHOULD NOT be terminating by default; terminating delegations SHOULD 
only be used when there is a compelling technical reason to do so.

A delegation for a subset of images MAY be a multi-role delegation. A multi-role

> delegation indicates that each of the delegatee roles MUST sign the same metadata.

Reviewer: ```suggestion
delegation indicates that each of the delegatee roles MUST agree on the same metadata (e.g., cryptographic hashes and file sizes) about images.
```

Delegations only apply to the Image repository. The Targets role on the Director
repository MUST NOT delegate metadata signing responsibility.

### **5.1.3. The Snapshot role {#snapshot_role}**

The Snapshot role SHALL produce and sign metadata about all Targets metadata

> the repository releases, including the current version number and hash of the main 

Reviewer: ```suggestion
the repository releases, including the current version number and hash of the top-level 
```

Targets metadata and the version numbers and hashes of all delegated targets
metadata, as described in {{snapshot_meta}}.

### **5.1.4. The Timestamp role {#timestamp_role}**

The Timestamp role SHALL produce and sign metadata indicating whether there are 
new metadata or images on the repository. It MUST do so by signing the metadata
about the Snapshot metadata file.

## **5.2. Metadata abstract syntax {#meta_syntax}**

### **5.2.1. Common Metadata Structures and Formats**

### **5.2.2. Root Metadata {#root_meta}**

### **5.2.3. Targets Metadata {#targets_meta}**

#### **5.2.3.1 Metadata about Images**

#### **5.2.3.2 Metadata about Delegations {#delegations_meta}**

### **5.2.4. Snapshot Metadata {#snapshot_meta}**

### **5.2.5 Timestamp Metadata {#timestamp_meta}**

## **5.3 Server / repository implementation requirements**

An Uptane implementation SHALL make the following services available to vehicles:

* Image repository
* Director repository
* Time server

### **5.3.1 Image Repository**

The Image repository exists to allow an OEM and/or its suppliers to upload images and
their associated metadata. It makes these images and their metadata available to 

> vehicles. The Image repository is designed to be primarily controlled by human actors,

Reviewer: Where should we discuss that the Image repository should use offline keys?

and updated relatively infrequently.

The Image repository SHALL expose an interface permitting the download of metadata
and images. This interface SHOULD be public.

The Image repository SHALL require authorization for writing metadata and images.

The Image repository SHALL provide a method for authorized users to upload images
and their associated metadata. It SHALL check that a user writing metadata and images
is authorized to do so for that specific image by checking the chain of delegations
for the image as described in {{delegations_meta}}.

The Image repository SHALL implement storage which permits authorized users to
write an image file using a unique filename, and later read the same file using
the same name. It MAY use any filesystem, key-value store, or database that
fulfills this requirement.

The Image repository MAY require authentication for read access.

### **5.3.2. Director Repository**

The Director repository instructs ECUs as to which images should be installed
by producing signed metadata on demand. Unlike the Image repository, it is

> mostly controlled by automated, online processes. It also consults a private

Reviewer: Where should we discuss that the Director repository primarily uses online keys?
